generalize U_matrix_kanamori and fix U_matrix doc

the-hampel opened the pull request:
* generalize U_matrix_kanamori to allow for Uprime != U-2J cases
* allow to generate full four index Uijkl tensor as in U_matrix_slater
* rename U_matrix to U_matrix_slater because a function should not have the same name a module (sphinx autodoc currently broken see emptiness of https://triqs.github.io/triqs/latest/documentation/python_api/triqs.operators.util.U_matrix.html )
* clean up weird before function comments and merge missing info into docstrings

Warning: this will break backward compatibility due to renaming U_matrix -> U_matrix_slater !!! Please discuss.

the-hampel committed:
generalize U_matrix_kanamori and fix U_matrix doc

* generalize U_matrix_kanamori to allow for Uprime != U-2J cases
* allow to generate full four index Uijkl tensor as in U_matrix_slater
* rename U_matrix to U_matrix_slater because a function should not have
  the same name a module (sphinx autodoc currently broken)
* clean up weird before function comments and merge missing info into
  docstrings

Wentzell commented:
@krivenko Could you have a look at this?

krivenko commented on python/triqs/operators/util/U_matrix.py:
```diff
@@ -18,17 +18,21 @@
-    Calculate the full four-index U matrix being given either radial_integrals or U_int and J_hund.
+    Calculate the full four-index U matrix
+
+    .. math:: U^{spherical}_{m1 m2 m3 m4} = \sum_{k=0}^{2l} F_k \alpha(l, k, m1, m2, m3, m4)
```
Since this equation is part of the docstring, it should be described what `F_k` and `\alpha(l, k, m1, m2, m3, m4)` stand for.

the-hampel replied:
addressed in e6725183

krivenko commented on python/triqs/operators/util/U_matrix.py:
```diff
@@ -131,43 +135,88 @@ def reduce_4index_to_2index(U_4index):
+    if not Up_int:
+        Up_int = U_int-2*J_hund
+    if not Jc_hund:
+        Jc_hund = Jc_hund
```
`Jc_hund = J_hund`

krivenko commented on python/triqs/operators/util/U_matrix.py:
```diff
@@ -131,43 +135,88 @@ def reduce_4index_to_2index(U_4index):
-        else:
-            U[m,mp]  = U_int - 3.0*J_hund
-            Uprime[m,mp] = U_int - 2.0*J_hund
+    if not Up_int:
```
`not Up_int` evaluates to `True` for `Up_int==0.0`. Is this the desired behavior?

the-hampel replied:
addressed in e6725183

krivenko commented on python/triqs/operators/util/U_matrix.py:
```diff
@@ -131,43 +135,88 @@ def reduce_4index_to_2index(U_4index):
-            Uprime[m,mp] = U_int - 2.0*J_hund
+    if not Up_int:
+        Up_int = U_int-2*J_hund
+    if not Jc_hund:
```
Same here.

krivenko commented on python/triqs/operators/util/U_matrix.py:
```diff
@@ -131,43 +135,88 @@ def reduce_4index_to_2index(U_4index):
+            if m == mp:
+                Uprime[m, mp] = U_int
+            else:
+                U[m, mp] = Up_int - 1.0*J_hund
```
`1.0*` looks a bit silly...

the-hampel replied:
addressed in e6725183

krivenko commented on python/triqs/operators/util/U_matrix.py:
```diff
@@ -131,43 +135,88 @@ def reduce_4index_to_2index(U_4index):
+
+    .. math:: U_{m m'}^{\sigma \bar{\sigma}} \equiv U_{m m' m m'}
+
+    If full_Uijkl=True is specified instead the full four index 
```
It would be nice to have a unit test that compares two Hamiltonians constructed from `U`/`Uprime` and from `Uijkl`.

the-hampel replied:
addressed in e6725183

the-hampel committed:
include @krivenko suggestions and add test

the-hampel commented:
Hi @Wentzell @krivenko , 
thank you for your suggestions. I fixed the flawed if statements, extended the docstrings, and added a unit test as suggested.

Wentzell reviewed (approved):
Thank you @the-hampel for this improvement and @krivenko for the feedback!
